### ISSUE TITLE
"no-useless-escape" reports warning on a regex pattern that contains non-omittable escape characters

### ISSUE THREAD

**Issue by SukkaW** [eslint/eslint#13826]
**Tell us about your environment**

* **ESLint Version:** ESLint 7.13.0 is affected
* **Node Version:** The issue can be reproduced on Node.js 10, 12, 14, 15
* **npm Version:** 6.14.8

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**

`@babel/eslint-parser`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
    "parser": "babel-eslint",
    "env": {
        "es6": true,
        "node": true
    },
    "extends": "eslint:recommended"
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
const regexp = /<[^<>\[\]]+?>/gm
console.log(regexp.match('<hey><[hey]>'))
```

**What did you expect to happen?**

There should be no issue detected with the code since the escape characters inside `/<[^<>\[\]]+?>/` cannot be omitted.

The code above should print:

```
["<hey>"]
```

**What actually happened? Please include the actual, raw output from ESLint.**

```
1:22 - Unnecessary escape character: \[. (no-useless-escape)
```

I have created [an ESLint demo](https://eslint.org/demo#eyJ0ZXh0IjoiY29uc3QgcmVnZXhwID0gLzxbXjw+XFxbXFxdXSs/Pi9nbVxuY29uc29sZS5sb2coJzxoZXk+=) for the issue.

As I mentioned before, the escape characters inside `regexp` cannot be omitted. If I omit the escape characters:

```js
const regexp = /<[^<>[]]+?>/gm
console.log('<hey><[hey]>'.match(regexp))
```

It will result in `null`.

**Comment by mdjermanovic** [comment]
Hi @SukkaW, thanks for the issue!

I believe this works as intended. The rule reports only `\[`, not `\]`.

[online demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVzZWxlc3MtZXNjYXBlOiBlcnJvciAqL1xuXG52YXIgcmVnZXhwID0gLzxbXjw+XFxbXFxdXSs/Pi9nbVxuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)

`\[` is indeed a useless escape in character classes. The original regex is equivalent to:

```js
const regexp = /<[^<>[\]]+?>/gm
```

**Comment by SukkaW** [comment]
@mdjermanovic Thanks for your explanation!

However, since the ESLint can't provide fix at the moment, should the issue to be converted to a feature request?

**Comment by mdjermanovic** [comment]
> However, since the ESLint can't provide fix at the moment, should the issue to be converted to a feature request?

Request to add autofixing? By design, this rule isn't auto-fixable because `\` that happens to be useless can represent different kinds of errors in the code, so we didn't want to assume that user always want to remove it (although it would produce a fully equivalent code).

So, instead of autofixing, this rule provides multiple [suggestions](https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions):

![image](https://user-images.githubusercontent.com/44349756/98689616-61d94080-236c-11eb-81d9-eb43d56d0094.png)

**Comment by nzakas** [comment]
Closing, as the rule is working as intended.
